### PR TITLE
TopNav: Dashboard settings 

### PR DESCRIPTION
**Pull request by torkelo**
Part of #50791

Ideally we should use a nested router for these pages (and make editIndex & editview URL paths), but opening dashboard settings needs to not loose any dashboard state (or panel edit state), nor URL state/variable state (current time range, variables values etc). I could not get a nested router to work in a way that preserved all state but I think that should be easier to pursue in separate PR. This already changes a lot and makes each of the settings views use a proper Page component (and the whole view is now not an full screen overlay like before but now just a normal page, the dashboard page is just hidden with css).

* [x] Make dashboard settings view not an overlay anymore (not absolutely positioned above everything), instead hide dashboard content and display settings view instead. 
* [x] Make edit index a URL parameter so we can use it outside each edit section to build the breadcrumbs (I know this is a bit messy and these should ideally be nested routes but that is a much bigger refactoring, especially to avoid unmounting the dashboard)
* [x] Make annotations section support deep breadcrumbs 
* [x] Make variables edit support breadcrumbs
* [x] Fix the double-page titles in many views 

For future PRS
* [ ] Make links edit support breadcrumbs
* [ ] Make versions compare support breadcrumbs

### CODE REVIEW

**Comment by torkelo** [comment]
Sorry for the massive PR, I realised a bit late that there was a way to not solve everything in one PR (So Links and Versions compare does not support breadcrumbs). But now all the other pages do work.

**Review by axelavargas — COMMENTED** [review]
Hi, @torkelo 👋🏾 great job with this PR (the settings pages have a much better user experience) 🥳

I was testing it out and found a couple of things :) (My tests were done with the `topnav` feature toggle `OFF`

#### Variables
1. "New" button is missing
Dash-settings-variables-MissingNewButton

2. Deleting a variable does not update the URL

Dash-Settings-Deleting-Var-Not-Updating


#### Annotations

1. Checkbox are misaligned

dash-settings-annotations-css-checkbox


#### Questions

Should we have "Dashboard" title section showing when the feature toggle is off?

Dash-Settings-should-be-here

**Review by ashharrison90 — COMMENTED** [review]
this looks great! 🙌 

feels like this section header should be the name of the dashboard (instead of just `Dashboard`) to match the breadcrumbs:
image

the dashboard settings -> permissions page has a trailing `;`:
image

**Comment by polibb** [comment]
Hi @torkelo 🙇‍♀️, the changes have made the experience of using the settings easier indeed. I tried hard-reloading of the page when I am on _DashboardName > Annotations > A Test Annotation_ and when I am on _DashboardName > Variables > A single Variable_. The URL loads correctly when reloading with `&editview=annotations&editIndex=0`, but it fails to load when trying to open `&editview=templating&editIndex=0`. 

It's interesting because it first loads the page and **then** it fails with:
<details> <summary> <code>
An unexpected error happened
DetailsError: Couldn't find variable with id:testavariable
</code> </summary>
<pre><code>
at ConnectFunction
at div
at div
at div
at Scrollbars 
at CustomScrollbar
at div
at div
at div
at Page
at VariableEditorContainerUnconnected
at ConnectFunction
at DashboardSettings 
at UnthemedDashboardPage 
at DashboardPage
at ConnectFunction 
at LoadableComponent 
at GrafanaRoute 
at Route
</code></pre> </details>

File: public/app/features/dashboard/components/DashboardSettings/DashboardSettings.tsx [hunk 1]
```diff
@@ -23,204 +21,205 @@ import { GeneralSettings } from './GeneralSettings';
...
-    height: 100%;
-    padding: 32px;
-    border: 1px solid ${theme.colors.border.weak};
-    background: ${theme.colors.background.primary};
-    border-radius: ${theme.shape.borderRadius()};
-  `,
-}));
+function getEditIndex(location: H.Location): number | undefined {
+  const editIndex = new URLSearchParams(location.search).get('editIndex');
+  if (editIndex != null) {
```
[the remark this answers is not in the document]
Comment (torkelo) [reply public/app/features/dashboard/components/DashboardSettings/DashboardSettings.tsx:221]: I think we use  `!=` heavily to check null & undefined

**Comment by torkelo** [comment]
@ashharrison90 

> feels like this section header should be the name of the dashboard (instead of just Dashboard) to match the breadcrumbs: 

Logically yes, but it somehow felt confusing.
* Dashboard names can be very short and related to the system it's about, or there is context needed with the folder, this can make the section heading very confusing (unless you know it's the dashboard title). 
* Dashboard names can be very long and won't fit as section heading (so would either wrap or ellipsis. 

Might still be the right decision but testing it felt off somehow, sections names should be more familiar and if we use the dashboard name there it's going to be very different each time the user enters dashboard settings. But if it always says "Dashboard" it will be instantly familiar that they are inside the dashboard settings section.

**Comment by ashharrison90** [comment]
sure makes sense 👍  doesn't have to hold this out but probably something to discuss with UX later

**Comment by torkelo** [comment]
> the changes have made the experience of using the settings easier indeed. I tried hard-reloading of the page when I am on DashboardName > Annotations > A Test Annotation and when I am on DashboardName > Variables > A single Variable. The URL loads correctly when reloading with &editview=annotations&editIndex=0, but it fails to load when trying to open &editview=templating&editIndex=0.

@polibb strange unable to replicate this.

**Review by axelavargas — COMMENTED** [review]
Hey @torkelo 👋🏾, I gave it a second try (again feature toggle off), and works as expected 🥳 , but I noticed there are some a11y issues regarding semantic HTML and Aria patterns in the `NavSection` component (this PR is using that component in the settings page)

Pa11y CI is already catching the error, it was not reported before because this component was behind the feature flag

You can check the error using `FastPass` :)

https://user-images.githubusercontent.com/239999/185084195-7d757f02-8cda-4568-a7c9-37c9327663cb.mp4

File: .betterer.results [hunk 2]
```diff
@@ -3874,10 +3874,26 @@ exports[`better eslint`] = {
     "public/app/features/dashboard/components/DashboardRow/DashboardRow.tsx:5381": [
       [0, 0, 0, "Unexpected any. Specify a different type.", "0"]
     ],
-    "public/app/features/dashboard/components/DashboardSettings/DashboardSettings.test.tsx:5381": [
+<<<<<<< HEAD
```
Comment (axelavargas) [inline .betterer.results:3877]: I think this was committed by mistake 😅

Comment (ashharrison90) [inline .betterer.results:3877]: think i'll have to look into how this passed CI 😬 

but definitely needs a `yarn betterer -u` committed + pushed

**Review by ashharrison90 — CHANGES_REQUESTED** [review]
needs the betterer conflict fixing + pushing (think just `yarn betterer -u` will do it), otherwise lgtm

i _think_ we can solve the a11y thing without reverting back to `div`s, but i don't wanna suggest that and be wrong and i don't want it to hold out this PR, so once it's in i'll have a play around and maybe raise a follow up 😄

**Review by axelavargas — APPROVED** [review]
🚀 work as expected (gave it a final test and checked the latest changes), there are no a11y issues 👏🏾 . I left a comment about a minor thing with betterer. 

ps, we might need to refactor `SectionNav` to[ be a11y compliant](https://www.deque.com/blog/a11y-support-series-part-1-aria-tab-panel-accessibility/), but we can definitely do that on a different PR

File: .betterer.results (hunk 2 above)
Comment (torkelo) [reply .betterer.results:3877]: oops

Comment (torkelo) [reply .betterer.results]: fixed

Comment (torkelo) [reply .betterer.results]: fixed

**Review by ashharrison90 — APPROVED** [review]
looks great! 🙌

**Pull request merged**